18-websockets/classworks/Anna_Volodkova/web/src/pages/Chat/Chat.jsx

```diff
@@ -0,0 +1,100 @@
+import React, {useState} from 'react';
+import './Chat.css';
+import Message from './components/Message';
+
+function Chat(props) {
+  const socket = props.socket;
+  const [value, setValue] = useState('');
+  const [userName, setUserName] = useState('');
+
+  const [isTyping, setIsTyping] = useState(false);
+  const [typingUser, setTypingUser] = useState('')
+
+  const [messages, setMessages] = useState([]);
+
+  socket.on('get-message', data => {
+    setMessages([...messages, data]);
+  });
+
+  if(isTyping){
+    socket.emit('send-isTyping', userName);
+  }
+
+  socket.on('user-typing', name => {
+    setTypingUser(name);
+  })
+
+
+
+  const sendMessage = () => {
+    if(!value) return;
+    socket.emit('send-message', value);
+  }
+
+  const sendUserName = () => {
+    if(!userName) return;
+    socket.emit('send-name', userName);
+  }
+
+  const sendTyping = () => {
+    if(!userName) return;
+    socket.emit('send-name', userName);
+  }
+
+  return (
+    <div className="chat">
+      <header>
+        <h1>Super Chat</h1>
+      </header>
+
+      <section>
+        <div id="change_username">
+          <input
+            id="username"
+            type="text"
+            value={userName}
+            onChange={event => setUserName(event.target.value)}
+          />
+          <button
+            id="send_username"
+            type="button"
+            onClick={sendUserName}
+          >
+            Change username
+          </button>
+        </div>
+      </section>
+
+      <section id="chatroom">
+        {
+          messages && messages.map( m => ( <Message username={m.username} message={m.value} />) )
+        }
+        <section id="feedback" />
+      </section>
+      {(typingUser) && <div>{typingUser} is typing</div>}
+      <section id="input_zone">
+        <input
+          id="message"
+          className="vertical-align"
+          type="text"
+          value={value}
+          onChange={event => {
+            setValue(event.target.value);
+            setIsTyping(true);
+            setTimeout(() => { setIsTyping(false); setTypingUser('')}, 1000)
```

Comment:
looks like, you will remove all typing users if the user doesn't type. I think typing users shouldn't be dependent on user actions.
I'd suggest adding a timeout for 2-4 secs after getting the typing event for deleting user. If you get a message about typing again you should restart timeout.